pkg/infra/iac2/templates_compiler.go

```diff
@@ -0,0 +1,189 @@
+package iac2
+
+import (
+	"embed"
+	_ "embed"
+	"fmt"
+	"github.com/klothoplatform/klotho/pkg/graph"
+	"github.com/klothoplatform/klotho/pkg/multierr"
+	"github.com/mitchellh/hashstructure/v2"
+	"github.com/pkg/errors"
+	"io"
+	"io/fs"
+	"reflect"
+	"regexp"
+	"sort"
+	"strings"
+)
+
+type (
+	varNamer interface {
+		VariableName() string
+	}
+
+	resource struct {
+		hash    string
+		element any
+	}
+
+	templatesCompiler struct {
+		templates             fs.FS
+		resourceGraph         *graph.Directed[resource]
+		resources             map[any]resource
+		templatesByStructName map[string]ResourceCreationTemplate
+	}
+)
+
+var (
+	//go:embed templates/*/factory.ts
+	standardTemplates embed.FS
+
+	nonIdentifierChars = regexp.MustCompile(`\W`)
+)
+
+func CreateTemplatesCompiler() *templatesCompiler {
+	subTemplates, err := fs.Sub(standardTemplates, "templates")
+	if err != nil {
+		panic(err) // unexpected, since standardTemplates is statically built into klotho
+	}
+	return &templatesCompiler{
+		templates:             subTemplates,
+		resourceGraph:         graph.NewDirected[resource](),
+		resources:             make(map[any]resource),
+		templatesByStructName: make(map[string]ResourceCreationTemplate),
+	}
+}
+
+func (tc templatesCompiler) AddResource(v any) {
+	if _, exists := tc.resources[v]; exists {
+		return
+
+	}
+	res := resource{
+		hash:    fmt.Sprintf(`%x`, len(tc.resources)),
+		element: v,
+	}
+	tc.resources[v] = res
+	tc.resourceGraph.AddVertex(res)
+	for _, child := range getStructValues(v) {
+		if reflect.TypeOf(child).Kind() == reflect.Struct {
+			tc.AddResource(child)
+			childRes := tc.getResource(child)
+			tc.resourceGraph.AddEdge(res, childRes)
+		}
+	}
+}
+
+func (tc templatesCompiler) getResource(v any) resource {
+	childRes, childExists := tc.resources[v]
+	if !childExists {
+		panic(fmt.Sprintf(`compiler has inconsistent state: no resource for %v`, v))
+	}
+	return childRes
+}
+
+func (tc templatesCompiler) RenderBody(out io.Writer) error {
+	// TODO: for now, assume a nice little tree
+	// TODO: need a stable sorting of outputs!
+
+	errs := multierr.Error{}
+	for _, res := range tc.resourceGraph.Roots() {
+		err := tc.renderResource(out, res.element)
+		errs.Append(err)
+	}
+	return errs.ErrOrNil()
+}
+
+func (tc templatesCompiler) RenderImports(out io.Writer) error {
+	// TODO: for now, assume a nice little tree
+
+	allImports := make(map[string]struct{})
+	for _, res := range tc.resources {
+		tmpl := tc.GetTemplate(res.element)
+		for statement, _ := range tmpl.imports {
+			allImports[statement] = struct{}{}
+		}
+	}
+
+	sortedImports := make([]string, 0, len(allImports))
+	for statement, _ := range allImports {
+		sortedImports = append(sortedImports, statement)
+	}
+
+	sort.Strings(sortedImports)
+	for _, statement := range sortedImports {
+		if _, err := out.Write([]byte(statement)); err != nil {
+			return err
+		}
+		if _, err := out.Write([]byte("\n")); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
```

Comment:
will this prevent dedupes? or is there any js logic we can reuse? I assume we want them at the top of the file, but im not quite sure if this will handle that